Simplenote/NotesListController.swift

```diff
@@ -0,0 +1,213 @@
+import Foundation
+import CoreData
+import SimplenoteFoundation
+import SimplenoteSearch
+
+
+// MARK: - NotesListController
+//
+class NotesListController: NSObject {
+
+    /// Core Data Kung Fu
+    ///
+    private let viewContext: NSManagedObjectContext
+    private lazy var notesController = ResultsController<Note>(viewContext: viewContext,
+                                                               matching: state.predicateForNotes(filter: filter),
+                                                               sortedBy: state.descriptorsForNotes(sortMode: sortMode))
+
+    /// FSM: Active State
+    ///
+    private(set) var state: NotesListState = .results {
+        didSet {
+            guard oldValue != state else {
+                return
+            }
+
+            refreshEverything()
+        }
+    }
+
+    /// Filter: Applied when we're not in Search Mode
+    ///
+    var filter: NotesListFilter = .everything {
+        didSet {
+            guard oldValue != filter else {
+                return
+            }
+
+            refreshPredicates()
+        }
+    }
+
+    /// Refreshes the Internal State, so that Search Results matching the specified Keyword will be filtered out.
+    /// - Important: It's up to the caller to invoke `performFetch`!!
+    ///
+    var searchKeyword: String? {
+        get {
+            guard case let .searching(keyword) = self.state else {
+                return nil
+            }
+
+            return keyword
+        }
+        set {
+            guard let keyword = newValue, !keyword.isEmpty else {
+                state = .results
+                return
+            }
+
+            state = .searching(keyword: keyword)
+        }
+    }
```

Comment:
(Discussed over Slack!!). **BEAUTIFUL idea**, following up in a PR aiming at the tail of this chain, in order to avoid breaking everything that comes up next!!